### ISSUE TITLE
High CPU usage on macOS profile/release app, memory tab in DevTools crashing

### ISSUE THREAD

**Issue by seanmc86** [flutter/devtools#2776]
## Steps to Reproduce

1. Run `flutter create testapp`.
2. Add a TextField or TextFormField to the main screen
3. Run app on macOS in profile mode.

**Expected results:**

Very low CPU usage for basic widgets where app state is not changing.

**Actual results:**

CPU usage increases to 6-10% simply when the window is in focus and the text field selected, and remains high in the background, despite nothing actively changing in the app.

This behaviour becomes exponentially worse in larger applications, even if simply adding a few network images into a listview. When trying to inspect memory usage in the Memory tab of DevTools, snapshots are automatically taken so often that the page results constantly reload, making it too difficult to try read any of the information and in some instances DevTools in the browser crashes due to this.

Building the same application in profile/release mode for windows does not exhibit the same high CPU usage.

<details>
  <summary>Logs</summary>

```
Analyzing testapp...                                                   
No issues found! (ran in 1.7s)
```

```
Flutter (Channel master, 1.27.0-2.0.pre.102, on macOS 11.2.1 20D74 darwin-x64, locale en-GB)
    • Flutter version 1.27.0-2.0.pre.102 at /Users/sean/SDK/flutter
    • Framework revision b2e05062b6 (2 weeks ago), 2021-02-15 03:56:03 -0500
    • Engine revision 1d537824d6
    • Dart version 2.13.0 (build 2.13.0-30.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/sean/Library/Android/sdk
    • Platform android-30, build-tools 30.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.4, Build version 12D4e
    • CocoaPods version 1.10.1

[✗] Chrome - develop for the web (Cannot find Chrome executable at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome)
    ! Cannot find Chrome. Try setting CHROME_EXECUTABLE to a Chrome executable.

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 50.0.1
    • Dart plugin version 193.7547
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] IntelliJ IDEA Community Edition (version 2020.2.3)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 51.0.3
    • Dart plugin version 202.8070

[✓] VS Code (version 1.53.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.20.0

[✓] Connected device (2 available)
    • iPhone SE (2nd generation) (mobile) • DED0779D-C824-4E09-AB35-2B6B478A5770 • ios        • com.apple.CoreSimulator.SimRuntime.iOS-14-4 (simulator)
    • macOS (desktop)                     • macos                                • darwin-x64 • macOS 11.2.1 20D74 darwin-x64
    ! Error: iPhone has recently restarted. Xcode will continue when iPhone is unlocked. (code -14)

! Doctor found issues in 1 category.
```

</details>

**Comment by seanmc86** [comment]
> (triage) The engine team is interpreting this as a "how devtools is displaying the results", rather than a bona fide memory problem, so I'm going to transfer this to the devtools repo. If we're wrong give it back and we'll take a closer look.

Fair enough regarding my latter issue on DevTools, but should there not be some continuation in the discussion here regarding the high CPU usage? This was measured by Activity Monitor in the OS itself, not in DevTools.

**Comment by jacob314** [comment]
Fyi @terrylucas looks like auto-snapshotting is kicking in when perhaps it shouldn't. I don't see anything in the sample app that should be using enough memory to auto-snapshot but maybe I'm missing something. Also appears that there is some issue parsing the snapshot data on 1.27.0-2.0.pre.102. It could be worth verifying the issue still reproduces and that this wasn't a build when the memory snapshot format was broken relative to what DevTools expected.

**Comment by seanmc86** [comment]
> Hi @seanmc86
> Thanks for filing the issue, I can see a memory spike when focusing in a TextFormField on the `dev` and `master`
> 
> <img alt="Screenshot 2021-03-03 at 7 52 15 PM" width="1206" src="https://user-images.githubusercontent.com/48603081/109819467-18ce3480-7c5a-11eb-8a6f-93ae251c57e0.png">
> 
> code sample
> flutter doctor -v
> > This behaviour becomes exponentially worse in larger applications, even if simply adding a few network images into a listview.
> 
> To better address the devtools crash can you please provide your `flutter run --profile -d macos --verbose` with the crash trace and a complete reproducible minimal code sample and steps to reproduce.
> Thank you

@TahaTesser given that this was transferred to devtools but you reproduced my original issue (high CPU usage when focusing a TextField/TextFormField), should there be a new issue filed in flutter issues with these same conclusions? This is still a massive problem for me, even on the latest stable.

**Comment by TahaTesser** [comment]
@jacob314 
Does this issue needs to be filed for high CPU usage in flutter/flutter or Is it just devtools providing inaccurate data to the system?
What do you recommend to the above comment?
Thank you

**Comment by seanmc86** [comment]
> @jacob314
> Does this issue needs to be filed for high CPU usage in flutter/flutter or Is it just devtools providing inaccurate data to the system?
> What do you recommend to the above comment?
> Thank you

@TahaTesser please can this issue be split and the TextField high CPU usage reproduction be filed as a Flutter issue again? This part in particular doesn't appear to be a question for the devtools guys to answer, especially given that my example shows high CPU usage on an OS level. It is currently a major blocker for my team!

**Comment by TahaTesser** [comment]
@seanmc86 
Sure, file the flutter issue in flutter/flutter

**Comment by srawlins** [comment]
Closing as obsolete: The Memory screen and its snapshotting architecture have since been completely redesigned and replaced.

If you still face this issue, please let us know.
